### PR TITLE
osd: replace ceph:atomic_t with std::atomic in osd module.

### PR DESCRIPTION
**Pull request by cxwshawn**
trying to use std more, and atomic_t easier to use and no need
t implement more interfaces, like operator++,++ etc.

Signed-off-by: Xiaowei Chen <chen.xiaowei@h3c.com>

### CODE REVIEW

File: src/osd/OSD.h [hunk 1]
```diff
@@ -1024,18 +1024,18 @@ class OSDService {
     NOT_STOPPING,
     PREPARING_TO_STOP,
     STOPPING };
-  atomic_t state;
+  atomic_uint state;
```
Comment (tchaikov) [inline src/osd/OSD.h]: atomic_int

**Comment by cxwshawn** [comment]
@tchaikov  please review again, thanks .

**Comment by gregsfortytwo** [comment]
I think @adamemerson or somebody in his group was working on a more comprehensive set of changes to our atomics. Maybe he has thoughts about how incrementally we want to handle this transition.

**Comment by adamemerson** [comment]
I didn't have a plan to do a big replacement all at once, though I can if people like. It would have a portability advantage since all we'd depend upon is C++ <atomic> support, rather than pulling in libatomic_ops (or falling back on the spinlock.)

It would also open up the opportunity to use looser memory models where appropriate (though I would want to consult with the people more knowledgeable in any given subsystem before relaxing things.)

**Comment by cxwshawn** [comment]
@adamemerson @gregsfortytwo  so what's the meaning ?  is  this PR meaningful ? I am also trying to stdanlize ceph code.

**Comment by liewegas** [comment]
@cxwshawn It means this is a step in the right direction, and nobody was working on it yet.  :)

**Comment by liewegas** [comment]
retest this please

File: src/osd/OSD.h [hunk 2]
```diff
@@ -1241,32 +1241,32 @@ class OSD : public Dispatcher,
   }
 
 private:
-  atomic_t state;
+  atomic_uint state;
```
Comment (tchaikov) [inline src/osd/OSD.h]: should use `atomic_int`, as the `state` here is actually .

**Comment by cxwshawn** [comment]
@tchaikov  :+1: ,  thanks for the review, fixed

File: src/osd/OSD.h [hunk 3]
```diff
@@ -796,7 +796,7 @@ class OSDService {
 
 private:
   /// throttle promotion attempts
-  atomic_t promote_probability_millis; ///< probability thousands. one word.
+  atomic_uint promote_probability_millis; ///< probability thousands. one word.
```
Comment (cbodley) [inline src/osd/OSD.h]: i'd prefer if we were explicit about the `std::` part, especially in headers - we shouldn't assume that other headers are `using namespace std;`

Comment (liewegas) [inline src/osd/OSD.h]: agreed on std::

**Comment by cxwshawn** [comment]
@cbodley thanks for your advice, @liewegas  have fixed, please review.

**Comment by liewegas** [comment]
lgtm

**Comment by liewegas** [comment]
valgrind uninit conditoin errors:
<pre>
  <kind>UninitCondition</kind>
  <what>Conditional jump or move depends on uninitialised value(s)</what>
  <stack>
    <frame>
      <ip>0xD249370</ip>
      <obj>/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.19</obj>
      <fn>std::ostreambuf_iterator&lt;char, std::char_traits&lt;char&gt; &gt; std::num_put&lt;char, std::ostreambuf_iterator&lt;char, std::char_traits&lt;char&gt; &gt; &gt;::_M_insert_int&lt;unsigned long&gt;(std::ostreambuf_iterator&lt;char, std::char_traits&lt;char&gt; &gt;, std::ios_base&amp;, char, unsigned lon
g) const</fn>
    </frame>
    <frame>
      <ip>0xD24960C</ip>
      <obj>/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.19</obj>
      <fn>std::num_put&lt;char, std::ostreambuf_iterator&lt;char, std::char_traits&lt;char&gt; &gt; &gt;::do_put(std::ostreambuf_iterator&lt;char, std::char_traits&lt;char&gt; &gt;, std::ios_base&amp;, char, unsigned long) const</fn>
    </frame>
    <frame>
      <ip>0xD2552CD</ip>
      <obj>/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.19</obj>
      <fn>std::ostream&amp; std::ostream::_M_insert&lt;unsigned long&gt;(unsigned long)</fn>
    </frame>
    <frame>
      <ip>0x465DE2</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>operator&lt;&lt;</fn>
      <dir>/usr/include/c++/4.8</dir>
      <file>ostream</file>
      <line>171</line>
    </frame>
    <frame>
      <ip>0x465DE2</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSDService::promote_throttle_recalibrate()</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.2.0-1558-g22ad94c/src/osd</dir>
      <file>OSD.cc</file>
      <line>629</line>
    </frame>
    <frame>
      <ip>0x4BEE9F</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSD::tick()</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.2.0-1558-g22ad94c/src/osd</dir>
      <file>OSD.cc</file>
      <line>4336</line>
    </frame>
    <frame>
      <ip>0x4C7228</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>Context::complete(int)</fn>
      <dir>/srv/autobuild-ceph/gitbuilder.git/build/out~/ceph-10.2.0-1558-g22ad94c/src/./include</dir>
      <file>Context.h</file>
      <line>64</line>
    </frame>
</pre>
and several others, all in promote_throttle_recalibrate.  Maybe teh atomic isn't getting initialized to 0?  the old atomic_t did that.
/a/yuriw-2016-06-01_08:07:30-rados-wip-yuri-testing-distro-basic-smithi/227666

**Comment by adamemerson** [comment]
Atomic versions of primitive types are not initialized. I'd recommend just sticking a ' = 0' after each in the class declaration.

(Actually this is my preference in most cases. If the class has a whole bunch of stuff that will be initialized to the same value regardless of the constructor called or parameters given, I think it's better style to set its default at declaration. That makes the initializer list in the constructor less noisy.)

**Comment by cxwshawn** [comment]
@liewegas sorry for the latency, maybe this is caused by osd_types.h member variable-not-initialized, have fixed, please try again, @adamemerson yeah, I have changed the uninited variable initiated in class declaration other than constructor initialize list.

File: src/osd/OSD.h [hunk 4]
```diff
@@ -796,7 +796,7 @@ class OSDService {
 
 private:
   /// throttle promotion attempts
-  atomic_t promote_probability_millis; ///< probability thousands. one word.
+  std::atomic_uint promote_probability_millis; ///< probability thousands. one word.
```
Comment (athanatos) [inline src/osd/OSD.h]: Should be initialized here.

File: src/osd/OSD.h [hunk 5]
```diff
@@ -840,9 +840,9 @@ class OSDService {
 
   // -- tids --
   // for ops i issue
-  atomic_t last_tid;
+  std::atomic_uint last_tid;
```
Comment (athanatos) [inline src/osd/OSD.h]: Should be initialized here.

File: src/osd/OSD.h [hunk 6]
```diff
@@ -1024,18 +1024,18 @@ class OSDService {
     NOT_STOPPING,
     PREPARING_TO_STOP,
     STOPPING };
-  atomic_t state;
+  std::atomic_int state;
```
Comment (athanatos) [inline src/osd/OSD.h]: Should be initialized here.

File: src/osd/OSD.h [hunk 7]
```diff
@@ -1241,32 +1241,32 @@ class OSD : public Dispatcher,
   }
 
 private:
-  atomic_t state;
+  std::atomic_int state;
```
Comment (athanatos) [inline src/osd/OSD.h]: Should be initialized here.

File: src/osd/PG.h [hunk 8]
```diff
@@ -33,7 +33,7 @@
 #include "osd_types.h"
 #include "include/buffer_fwd.h"
 #include "include/xlist.h"
-#include "include/atomic.h"
+// #include "include/atomic.h"
```
Comment (athanatos) [inline src/osd/PG.h]: Remove this completely.

File: src/osd/PG.h [hunk 9]
```diff
@@ -245,7 +246,7 @@ class PG : DoutPrefixProvider {
    * put_unlock() when done with the current pointer (_most common_).
    */  
   mutable Mutex _lock;
-  atomic_t ref;
+  std::atomic_uint ref;
```
Comment (athanatos) [inline src/osd/PG.h]: Should be initialized here.

File: src/osd/osd_types.h [hunk 10]
```diff
@@ -4280,24 +4281,26 @@ enum scrub_error_type {
 // PromoteCounter
 
 struct PromoteCounter {
-  atomic64_t attempts, objects, bytes;
+  std::atomic_ullong  attempts{0};
```
Comment (athanatos) [inline src/osd/osd_types.h]: Yep, like this.

File: src/osd/osd_types.h [hunk 11]
```diff
@@ -4280,24 +4281,26 @@ enum scrub_error_type {
...
+  std::atomic_ullong  objects{0};
```
Comment (athanatos) [inline src/osd/osd_types.h]: Yep, like this.

File: src/osd/osd_types.h [hunk 12]
```diff
@@ -4280,24 +4281,26 @@ enum scrub_error_type {
...
+  std::atomic_ullong  bytes{0};
```
Comment (athanatos) [inline src/osd/osd_types.h]: Yep, like this.

**Comment by athanatos** [comment]
You seem to have missed several more initializations.  We'll re-add needs-qa once those are fixed.

**Comment by cxwshawn** [comment]
@athanatos  since the missed you said has been initialized by using constructor initialize list, so is there any need to reinitialize using xxx a{0} ?

**Comment by athanatos** [comment]
You are claiming that they are already initialized in the constructor list?  I'd actually prefer you moved the initialization to the declaration like you did with the other variables.  Are those not also initialized in the constructor?

**Comment by cxwshawn** [comment]
@athanatos yes, exactly, the former variables has been initialized in the constructor list, so I think there is no need to reinitialize, but since declaration initialization is preferred, I will change all the atomic variables to this form.

**Comment by cxwshawn** [comment]
@athanatos all modified atomic var has been initialized when declaration, please help merge, thanks!

**Comment by athanatos** [comment]
make check seems to be failing

**Comment by cxwshawn** [comment]
@athanatos it's a little weired, does this failure relate with this change  ?

**Comment by tchaikov** [comment]
@cxwshawn you might want to rebase your changeset against master for a green build.

**Comment by cxwshawn** [comment]
@tchaikov @athanatos rebased, but still failed, is there any hint for this failure ? I cannot see any information related with this change.

**Comment by tchaikov** [comment]
tested in http://pulpito.ceph.com/kchai-2016-07-10_07:48:29-rados-wip-kefu-testing---basic-mira/ 

the non-environmental failures are addressed by #10234 .

**Pull request merged**